include/swift/SIL/SILType.h

```diff
@@ -553,33 +547,346 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS, SILType T) {
+  }
+};
+
+/// Transient wrapper for SIL-level argument conventions. This abstraction helps
+/// handle the transition from canonical SIL conventions to lowered SIL
+/// conventions.
+class SILFunctionConventions {
+public:
+  SILModuleConventions silConv;
+  CanSILFunctionType funcTy;
+
+  SILFunctionConventions(CanSILFunctionType funcTy, SILModule &M)
+      : silConv(M), funcTy(funcTy) {}
+
+  SILFunctionConventions(CanSILFunctionType funcTy,
+                         SILModuleConventions silConv)
+      : silConv(silConv), funcTy(funcTy) {}
+
+  //===--------------------------------------------------------------------===//
+  // SILModuleConventions API for convenience.
+  //===--------------------------------------------------------------------===//
+
+  bool isSILIndirect(SILParameterInfo param) const {
+    return silConv.isSILIndirect(param);
+  }
+
+  bool isSILIndirect(SILResultInfo result) const {
+    return silConv.isSILIndirect(result);
+  }
+
+  SILType getSILType(SILParameterInfo param) const {
+    return silConv.getSILType(param);
+  }
+
+  SILType getSILType(SILResultInfo result) const {
+    return silConv.getSILType(result);
+  }
+
+  //===--------------------------------------------------------------------===//
+  // SIL results.
+  //===--------------------------------------------------------------------===//
+
+  /// Get the normal result type of an apply that calls this function.
+  /// This does not include indirect SIL results.
+  SILType getSILResultType() {
+    if (silConv.loweredAddresses)
+      return funcTy->getDirectFormalResultsType();
+
+    return funcTy->getAllResultsType();
+  }
+
+  /// Get the SIL type for the single result which may be direct or indirect.
+  SILType getSingleSILResultType() {
+    return getSILType(funcTy->getSingleResult());
+  }
+
+  /// Get the error result type.
+  SILType getSILErrorType() { return getSILType(funcTy->getErrorResult()); }
+
+  /// Get the number of SIL results passed as address-typed arguments.
+  unsigned getNumIndirectSILResults() const {
+    return silConv.loweredAddresses ? funcTy->getNumIndirectFormalResults() : 0;
+  }
+
+  /// Are any SIL results passed as address-typed arguments?
+  bool hasIndirectSILResults() const { return getNumIndirectSILResults() != 0; }
+
+  using IndirectSILResultIter = SILFunctionType::IndirectFormalResultIter;
+  using IndirectSILResultRange = SILFunctionType::IndirectFormalResultRange;
+
+  /// Return a range of indirect result information for results passed as
+  /// address-typed SIL arguments.
+  IndirectSILResultRange getIndirectSILResults() const {
+    if (silConv.loweredAddresses)
+      return funcTy->getIndirectFormalResults();
+
+    auto filter = llvm::make_filter_range(
+        makeIteratorRange((const SILResultInfo *)0, (const SILResultInfo *)0),
+        SILFunctionType::IndirectFormalResultFilter());
+    return makeIteratorRange(filter.begin(), filter.end());
+  }
+
+  struct SILResultTypeFunc {
```

Comment:
Should uses ever see this? Maybe hide it in a detail struct? Not wedded to this, just a thought.